### ISSUE TITLE
Ability to ignore globals created by window?

### ISSUE THREAD

**Issue by Lexicality** [eslint/eslint#3959]
There are a lot of silly globals produced by everything in window being also global, such as `name`, `self` and `top`.
The code that I'm working on uses these words in it, and this is making my refactor annoyingly difficult.
For instance I can't use the `builtinGlobals` option for `no-shadow` and `no-undef` just now didn't notice when I deleted a variable called `self` without removing all uses first.
Is there a way for me to make ESLint not know about all the implicit variables without disabling the browser env and manually adding `window` and `document` to the global list?

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by Lexicality** [comment]
version:
```
$ eslint -v
v1.5.0
```
.eslintrc:
```
rules:
  indent:
    - 2
    - 4
  # quotes:
  #   - 1
  #   - single
  linebreak-style:
    - 2
    - unix
  semi:
    - 2
    - always
  no-console: 1
  no-debugger: 1
  no-empty: 1
  valid-typeof: 1
  accessor-pairs: 1
  block-scoped-var: 2
  consistent-return: 2
  curly: 1
  dot-location:
    - 2
    - property
  eqeqeq:
    - 1
    - "allow-null"
  guard-for-in: 1
  no-alert: 1
  no-else-return: 1
  no-div-regex: 1
  no-eval: 2
  no-extend-native: 2
  no-extra-bind: 1
  no-fallthrough: 1
  no-floating-decimal: 1
  no-implied-eval: 2
  no-invalid-this: 0 # Currently the way we do things angers this rule
  no-lone-blocks: 2
  no-loop-func: 2
  no-multi-spaces: 1
  no-multi-str: 2
  no-new-func: 2
  no-new-wrappers: 2
  no-new: 1
  no-octal-escape: 2
  no-proto: 1
  no-redeclare:
    - 2
    - builtinGlobals: true
  no-return-assign: 1
  no-script-url: 2
  no-self-compare: 2
  no-sequences: 1
  no-throw-literal: 2
  no-unused-expressions: 2
  no-useless-concat: 1
  no-void: 2
  no-warning-comments:
    - 1
    -
      terms:
        - todo
        - fixme
        - hack
        - bug
        - why
      location: "start"
  no-with: 2
  vars-on-top: 0
  wrap-iife:
    - 2
    - inside
  yoda:
    - 1
    - never
  strict:
    - 2
    - global
  no-label-var: 2
  no-shadow-restricted-names: 2
  no-shadow:
    - 1
    -
      builtinGlobals: false # There are a lot of stupid globals like 'name'
      hoist: "all"
  no-undef-init: 2
  no-unused-vars:
    - 1
    -
      vars: "all"
      args: "none"
  no-use-before-define: 2
  new-parens: 2
  no-array-constructor: 2
  no-new-object: 2
  no-unneeded-ternary: 2
  arrow-parens: 2
  constructor-super: 2
  no-class-assign: 2
  no-const-assign: 2
  no-dupe-class-members: 2
  no-this-before-super: 2
  require-yield: 1
  # Non-subjective
  eol-last: 2
  no-trailing-spaces: 2
  # Let's get subjective!
  array-bracket-spacing: 1
  block-spacing: 1
  brace-style:
    - 1
    - 1tbs
  # camelcase: 1
  comma-spacing: 1
  comma-style:
    - 1
    - last
  computed-property-spacing: 1
  consistent-this:
    - 1
    - self
  func-style:
    - 1
    - declaration
  jsx-quotes: 1
  key-spacing:
    - 1
    -
      beforeColon: false
      afterColon: true
      mode: "minimum"
      # align: "value" # Some Day :(
  lines-around-comment:
    - 1
    -
      beforeBlockComment: true
      afterBlockComment: false
      beforeLineComment: false
      afterLineComment: false
      allowBlockStart: true
      allowBlockEnd: false
  max-nested-callbacks:
    - 1
    - 3
  new-cap: 1
  no-inline-comments: 1 # These break formatters (yes this comment is ironic)
  no-lonely-if: 1
  no-mixed-spaces-and-tabs:
    - 1
    - smart-tabs
  no-multiple-empty-lines:
    - 1
    - max: 3
  no-nested-ternary: 1
  no-spaced-func: 1
  padded-blocks:
    - 1
    - never
  quote-props:
    - 1
    - consistent-as-needed
  semi-spacing: 1
  operator-linebreak: 1
  object-curly-spacing:
    - 1
    - always
  one-var:
    - 1
    - initialized: "never"
  space-after-keywords: 1
  space-before-keywords: 1
  space-before-blocks: 1
  space-before-function-paren:
    - 1
    - never
  space-in-parens:
    - 1
    - never
  space-infix-ops: 1
  space-return-throw-case: 1
  space-unary-ops: 1
  spaced-comment:
    - 1
    - always
    -
      markers:
        - "/"
        - "!"
      exceptions:
        - "-"
        - "*"
  arrow-spacing: 1
  generator-star-spacing: 1
  # prefer-arrow-callback: 1
  comma-dangle:
    - 1
    - "never"
env:
  es6: true
  browser: true
extends: 'eslint:recommended'
ecmaFeatures:
  jsx: true
  modules: true
  experimentalObjectRestSpread: true
plugins:
  - react
```

**Comment by mysticatea** [comment]
If there is an option that `no-shadow` and `no-undef` have ignoring specific globals, it's useful. I want it.
On the other hand, `no-redeclare` should not ignore these.

**Comment by nzakas** [comment]
@Lexicality can you explain what you're saying from ESLint? I'm not understanding your use case.

**Comment by mysticatea** [comment]
Although I think what I will say in this comment is too peculiar as the core rule, I try to say.

----

The `browser` environment has many variables of general names: `name`, `top`, `event`, ...

This can make ESLint missing detection of typo.

```js
/* eslint-env browser */
/* eslint no-undef: 2, no-unused-vars: 2 */
function foo(nama) {
    // one process
    if (nama === "foo") { }

    // other process...
    if (name === "bar") { }
}
```

This pattern, `nama` and `name` are not warned.
Before, I had sometimes gotten a mistake while I was renaming variables for refactoring because of this (and the `typeof` exception of `no-undef`).

On the other hand, that `no-redeclare` reports variables of the same name as built-in globals is helpful.

```js
/* eslint-env browser */
var top = 0;
var left = 0;

console.log(top); // [object Window]
```

So, just removing global variables of general names is not good.

From the above, I thought an exception that `no-undef` and `no-shadow` ignore specific global variables may be useful (the global variables are not used often).

```js
{
    "no-undef": [2, {"ignoreBuiltinGlobals": ["name", "top", "event", "Buffer"]}],
    "no-redeclare": [2, {"builtinGlobals": true}],
    "no-shadow": [2, {"builtinGlobals": true,  "ignoreBuiltinGlobals": ["name", "top", "event", "Buffer"]}]
}
```

Maybe, this is related in #3966.

**Comment by nzakas** [comment]
We can't really talk about solutions until @Lexicality defines a problem. Right now we don't have enough detail to know what the real issue is.

**Comment by Lexicality** [comment]
Oh sorry, lost this in my inbox. @mysticatea described the problem exactly - the ESLint assumes that any action on a variable `name` is kosher because `window.name` exists, regardless of what is in scope at the time.

**Comment by nzakas** [comment]
What does "is kosher" mean to you? What rules are you talking about? Give us a specific problem scenario so we know what you're talking about.

**Comment by Lexicality** [comment]
```javascript
/* eslint-env browser */
/* eslint no-undef: 2, no-unused-vars: 2 */
function foo(nama) {
    // one process
    if (nama === "foo") { }

    // other process...
    if (name === "bar") { } // <--- This is not an error, but should be
}
```

**Comment by nzakas** [comment]
This is behaving as expected. You can turn off the browser environment if you want to manually define the globals you want to allow.

**Comment by Lexicality** [comment]
I suppose that is a solution, though not one I'm very happy about.
I did a bit of research and I'm now not sure what to think.
On the one hand I would consider either of these to be a programming error:
```javascript
function foo() {
    name = "foo";
}

function bar() {
    var bar = top;
    // do stuff with bar
}
```
but I wouldn't consider either of these to be:
```javascript
function foo() {
    console.log("foo")
}

function bar() {
    localStorage.set("bar", 5)
}
```
but both examples are using globals magically lifted from the window object.

If I turn off the browser environment then `no-redeclare`'s `builtinGlobals` becomes less useful as it doesn't catch this trap:

```javascript
  var top = 0;
  var left = 0;
  console.log(top + " " + left); // prints "[object Window] 0"
  console.log(top * left); // prints "NaN"
```

But also that doesn't ~really~ affect me because we use ES6 modules and they don't write to the global state.

Hmm.
![batman looking pensive](https://media0.giphy.com/media/a5viI92PAF89q/200_s.gif "batman looking pensive")

**Comment by nzakas** [comment]
Yeah, I'm sorry, I'm just not sure we can meet your expectations in this regard. I'd suggest using your own list of globals.
